src/Shared/TagTransformerJsonHelper.cs

```diff
@@ -0,0 +1,192 @@
+            byte[] byteArray => array.GetType() == typeof(sbyte[]) ? WriteToString((sbyte[])array, WriteSByteValue, 3) : WriteToString(byteArray, WriteByteValue, 3),
+            short[] shortArray => array.GetType() == typeof(ushort[]) ? WriteToString((ushort[])array, WriteUShortValue, 5) : WriteToString(shortArray, WriteShortValue, 5),
+            int[] intArray => array.GetType() == typeof(uint[]) ? WriteToString((uint[])array, WriteUIntValue, 10) : WriteToString(intArray, WriteIntValue, 10),
+            long[] longArray => array.GetType() == typeof(ulong[]) ? WriteToString((ulong[])array, WriteULongValue, 20) : WriteToString(longArray, WriteLongValue, 20),
+
+            float[] floatArray => WriteToString(floatArray, WriteFloatValue, 15),
+            double[] doubleArray => WriteToString(doubleArray, WriteDoubleValue, 23),
+
+            _ => throw new NotSupportedException($"Unexpected serialization of array of type {array.GetType()}."),
+        };
+
+    private static void WriteCharValue(char v, Utf8JsonWriter writer)
+    {
+        Span<char> charSpan = stackalloc char[1];
+        charSpan[0] = v;
+        writer.WriteStringValue(charSpan);
+    }
+
+    private static void WriteStringValue(string v, Utf8JsonWriter writer) => writer.WriteStringValue(v);
+
+    private static void WriteBooleanValue(bool v, Utf8JsonWriter writer) => writer.WriteBooleanValue(v);
+
+    private static void WriteByteValue(byte v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteSByteValue(sbyte v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteShortValue(short v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteUShortValue(ushort v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteIntValue(int v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteUIntValue(uint v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteLongValue(long v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteULongValue(ulong v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteFloatValue(float v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteDoubleValue(double v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static string WriteToString<T>(T[] data, Action<T, Utf8JsonWriter> writeAction, int elementSizeHint)
+    {
+        // Some rough estimate of the necessary size. We multiple the length of the array (+1 to be on the safe side)
+        // by the estimated size of an element and add 2 for the brackets.
+        using PooledByteBufferWriter bufferWriter = new PooledByteBufferWriter(((data.Length + 1) * (elementSizeHint + 1)) + 2);
+        using (Utf8JsonWriter writer = new Utf8JsonWriter(bufferWriter))
```

Comment:
I'm curious to learn the reasons on choosing `PooledByteBufferWriter` over the `Stream` overload?
https://github.com/open-telemetry/opentelemetry-dotnet/blob/e7f7cd68759d81aa3a34cb88d1b2590e09c21e89/src/OpenTelemetry.Exporter.Zipkin/ZipkinExporter.cs#L241C21-L241C62

Reply:
Well `Stream` itself is just an interface (abstract class, but that's the same thing basically) - the interesting bit is where does the implementation of it gets the memory to store the results in. In this case we need to store the results for only a brief moment, then convert it to string and drop it.
The typical implementation would use `MemoryStream`, but that will allocate twice for each serialization (the `MemoryStream` itself has to be allocated and then we have to allocate the buffer within it). The `PooledByteBufferWriter` avoids the second allocation (the buffer is taken from an `ArrayPool`). So less stress on GC.
We could also implemented something like `PooledByBufferStream` and use the `Stream` overload, but there would be basically no difference in behavior - and the `Stream` is more complex to implement (more methods to implement and the writing is more complex) and probably less performant as the JSON writer can't write directly to the internal buffer - so it would require additional memory copy.

Unfortunately we can't use the same trick in the Zipkin exporter as it has to implement `HttpContent` APIs which are all `Stream` based.